dvc/remote/local/__init__.py

```diff
@@ -569,3 +570,52 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum, force):
+        # Relinking with `symlink/hardlink/reflink` is cheap, but if we need `copy` links, then to save
+        # on copying we should only relink if we already have `hardlink` or a `symlink`, we don't care
+        # about reflinks, because they are indistinguishable from copy anyway.
+        return not (
+            not force
+            and self.exists(path_info)
+            and self.state.get(path_info) == checksum
+            and self._has_same_protection_as_cache(path_info)
+            and not System.is_hardlink(path_info)
+            and not System.is_symlink(path_info)
+            and self._is_cache_type_copy()
+        )
+
+    def _is_cache_type_copy(self):
+
+        if self.cache_types[0] == "copy":
+            return True
+
+        if not self.path_info:
+            return False
+
+        link_test_dir = self.path_info / "tmp_test"
+        try:
+            self.makedirs(link_test_dir)
+
+            file1 = link_test_dir / uuid()
```

Comment:
So you are testing linking within a single dir, which wrong. What you should really be testing is if you are able to link from cache_dir to the workspace, as we've discussed, because they might be on different fs.

Reply:
good point, ill refactor